### PR TITLE
Refactors the Engine Interface

### PR DESCRIPTION
**Pull request by hsubbaraj-spiral**
## Describe your changes and why you are making these changes
This introduces the Engine interface that sits between our handlers and execution engines. We create an implementation aqEngine that will handle orchestration for aqueduct-orchestrated engines. This PR unblocks the implementations of the K8s and Lambda integrations that are the next steps of the project.

Changes to note:

- [x] Orchestrator -> Engine.
- [x] Decoupling the workflow execution from the executor. (Now handled by ScheduleWorkflow() and ExecuteWorkflow())
- [x] Introducing a CronjobManager interface that Engine uses.
- [x] Refactors PreviewHandler to push most of the logic into aqEngine.PreviewWorkflow().

What this PR doesn't do:

- [ ] [Entirely remove the executor](https://linear.app/aqueducthq/issue/ENG-1444/remove-executor)
- [ ] [Remove cronjob handling from the JobManager](https://linear.app/aqueducthq/issue/ENG-1482/remove-cronjob-management-from-jobmanager)
- [ ] [Move JobManager creation from Server to Handlers](https://linear.app/aqueducthq/issue/ENG-1385/refactor-orchestrator-interface)
- [ ] Edit the data model to include engine config (this will be introduced by [Saurav's PR](https://github.com/aqueducthq/aqueduct/pull/228/files))

## Related issue number (if any)
https://linear.app/aqueducthq/issue/ENG-1385/refactor-orchestrator-interface

## Checklist before requesting a review
- [x] I have created a descriptive PR title. The PR title should complete the sentence "This PR...".
- [x] I have performed a self-review of my code.
- [ ] I have included a small demo of the changes. For the UI, this would be a screenshot or a Loom video.
- [ ] If this is a new feature, I have added unit tests and integration tests.
- [x] I have run the integration tests locally and they are passing.
- [x] I have run the linter script locally (See `python3 scripts/run_linters.py -h` for usage).
- [x] All features on the UI continue to work correctly.
- [ ] Added one of the following CI labels:
    - `run_integration_test`: Runs integration tests
    - `skip_integration_test`: Skips integration tests (Should be used when changes are ONLY documentation/UI)

### CODE REVIEW

File: src/golang/cmd/server/server/aqueduct_server.go [hunk 1]
```diff
@@ -19,9 +19,10 @@ import (
 	"github.com/aqueducthq/aqueduct/lib/job"
 	"github.com/aqueducthq/aqueduct/lib/logging"
 	"github.com/aqueducthq/aqueduct/lib/vault"
+	"github.com/aqueducthq/aqueduct/lib/workflow/engine"
 	"github.com/aqueducthq/aqueduct/lib/workflow/operator/connector/github"
 	"github.com/dropbox/godropbox/errors"
-	"github.com/go-chi/chi/v5"
+	"github.com/go-chi/chi"
```
Comment (likawind) [inline src/golang/cmd/server/server/aqueduct_server.go:25]: this seems to be a bad merge with #260 ?

File: src/golang/cmd/server/server/cron.go [hunk 2]
```diff
@@ -101,14 +105,14 @@ func (s *AqServer) initializeWorkflowCronJobs(ctx context.Context) error {
 			if wf.Schedule.Paused {
 				wf.Schedule.CronSchedule = ""
 			}
+			name := shared_utils.AppendPrefix(wf.Id.String())
```
Comment (likawind) [inline src/golang/cmd/server/server/cron.go:108]: this reads directly from DB. Do we need any sync from engine here?

File: src/golang/cmd/server/server/database.go [hunk 3]
```diff
@@ -215,3 +216,32 @@ func CreateWriters(dbConfig *database.DatabaseConfig) (*Writers, error) {
 		WorkflowDagResultWriter: workflowDagResultWriter,
 	}, nil
 }
+
+func GetEngineReaders(readers *Readers) *engine.EngineReaders {
```
Comment (likawind) [inline src/golang/cmd/server/server/database.go:220]: there seems to be some overlaps between engine readers and server readers. Curious about what's the difference and how should the caller decide which one to use?

File: src/golang/lib/workflow/engine/aq_engine.go [hunk 4]
```diff
@@ -0,0 +1,811 @@
...
+	WorkflowWatcherWriter   workflow_watcher.Writer
+	OperatorWriter          operator_db.Writer
+	OperatorResultWriter    operator_result.Writer
+	ArtifactWriter          artifact_db.Writer
+	ArtifactResultWriter    artifact_result.Writer
+	NotificationWriter      notification.Writer
+}
+
+type aqEngine struct {
+	database       database.Database
```
Comment (likawind) [inline src/golang/lib/workflow/engine/aq_engine.go:74]: I would consider using public fields. It's good enough to control access using top-level private type.  One past experience is #234 . There are two benefits both reflected in the mentioned PR:
* easier to dump the entire struct to json string for future logging and debugging. Json marshal will hide private fields.
* It's easier to move things around before codebase get stable. There's less renaming you would need to worry when one day you suddenly want to expose some field

**Review by likawind — COMMENTED** [review]
Looks good overall! I assume everything besides interface implementation are moving-arounds, which doesn't change any application logics. Besides that, just a few minor comments and questions

File: src/golang/cmd/server/handler/preview.go [hunk 5]
```diff
@@ -153,128 +120,36 @@ func (h *PreviewHandler) Perform(ctx context.Context, interfaceArgs interface{})
...
-		return errorRespPtr, http.StatusInternalServerError, errors.Wrap(err, "Error creating orchestrator.")
-	}
-
-	defer orch.Finish(ctx)
-	status, err := orch.Execute(ctx)
-	if err != nil && err != orchestrator.ErrOpExecSystemFailure && err != orchestrator.ErrOpExecBlockingUserFailure {
+	if err != nil && err != engine.ErrOpExecSystemFailure && err != engine.ErrOpExecBlockingUserFailure {
 		return errorRespPtr, http.StatusInternalServerError, errors.Wrap(err, "Error executing the workflow.")
 	}
+	log.Info(workflowPreviewResult)
```
Comment (saurav-c) [inline src/golang/cmd/server/handler/preview.go:137]: nit: remove print

File: src/golang/cmd/server/handler/refresh_workflow.go [hunk 6]
```diff
@@ -28,11 +29,14 @@ type RefreshWorkflowArgs struct {
...
 
-	Database       database.Database
-	JobManager     job.JobManager
-	GithubManager  github.Manager
-	Vault          vault.Vault
-	WorkflowReader workflow.Reader
+	Database database.Database
+	Engine   engine.Engine
+
+	ArtifactReader        artifact.Reader
```
Comment (saurav-c) [inline src/golang/cmd/server/handler/refresh_workflow.go:35]: Where are all of these readers used?

File: src/golang/cmd/server/handler/register_workflow.go [hunk 7]
```diff
@@ -46,23 +52,31 @@ type RegisterWorkflowHandler struct {
...
-
-	ArtifactWriter        artifact.Writer
-	OperatorWriter        operator.Writer
-	WorkflowWriter        workflow.Writer
-	WorkflowDagWriter     workflow_dag.Writer
-	WorkflowDagEdgeWriter workflow_dag_edge.Writer
-	WorkflowWatcherWriter workflow_watcher.Writer
+	Engine        engine.Engine
+
+	ArtifactReader        artifact.Reader
```
Comment (saurav-c) [inline src/golang/cmd/server/handler/register_workflow.go:57]: Are these readers and writers used directly, or are they part of `Engine`?

File: src/golang/go.mod [hunk 8]
```diff
@@ -6,6 +6,7 @@ require (
 	github.com/apache/airflow-client-go/airflow v0.0.0-20220509204651-4f1b26e4a5d0
 	github.com/aws/aws-sdk-go v1.40.33
 	github.com/dropbox/godropbox v0.0.0-20200228041828-52ad444d3502
+	github.com/go-chi/chi v1.5.4
```
Comment (saurav-c) [inline src/golang/go.mod:9]: Will this cause a conflict based on the bug you were seeing earlier?

File: src/golang/go.sum [hunk 9]
```diff
@@ -63,6 +63,8 @@ github.com/envoyproxy/go-control-plane v0.9.1-0.20191026205805-5f8ba28d4473/go.m
 github.com/envoyproxy/go-control-plane v0.9.4/go.mod h1:6rpuAdCZL397s3pYoYcLgu1mIlRU8Am5FuJP05cCM98=
 github.com/envoyproxy/go-control-plane v0.9.9-0.20201210154907-fd9021fe5dad/go.mod h1:cXg6YxExXjJnVBQHBLXeUAgxn2UodCpnH306RInaBQk=
 github.com/envoyproxy/protoc-gen-validate v0.1.0/go.mod h1:iSmxcyjqTsJpI2R4NaDN7+kN2VEUnK/pcBlmesArF7c=
+github.com/go-chi/chi v1.5.4 h1:QHdzF2szwjqVV4wmByUnTcsbIg7UGaQ0tPF2t5GcAIs=
```
Comment (saurav-c) [inline src/golang/go.sum:66]: Related to above

File: src/golang/lib/cronjob/process.go [hunk 10]
```diff
@@ -0,0 +1,134 @@
...
+	"time"
+
+	"github.com/dropbox/godropbox/errors"
+	"github.com/go-co-op/gocron"
+)
+
+type cronMetadata struct {
+	// If the cronJob is nil, it means the corresponding workflow has been paused.
+	cronJob *gocron.Job
+	// We need to store the job spec because when the workflow is resumed from the pause state, we
```
Comment (saurav-c) [inline src/golang/lib/cronjob/process.go:15]: nit: Should this be removed?

**Review by saurav-c — COMMENTED** [review]
Dropped some comments, generally looks good!

File: src/golang/cmd/executor/executor/executor.go [hunk 11]
```diff
@@ -20,17 +20,6 @@ type Executor interface {
...
-		workflowSpec, ok := spec.(*job.WorkflowSpec)
-		if !ok {
-			return nil, job.ErrInvalidJobSpec
-		}
-
-		base, err := NewBaseExecutor(workflowSpec.ExecutorConfig)
-		if err != nil {
-			return nil, err
-		}
-		return NewWorkflowExecutor(workflowSpec, base)
```
Comment (kenxu95) [inline src/golang/cmd/executor/executor/executor.go:33]: should we remove this case from the process job manager too?

File: src/golang/lib/workflow/engine/engine.go [hunk 12]
```diff
@@ -0,0 +1,71 @@
...
+}
+
+type SelfOrchestratedEngine interface {
+	Engine
+
+	SyncWorkflow(
+		ctx context.Context,
+		dbWorkflowDag *workflow_dag.DBWorkflowDag,
+	)
+}
```
Comment (kenxu95) [inline src/golang/lib/workflow/engine/engine.go:71]: Can we add some docstrings on each of these interfaces indicating 1) what scenarios each should be used for and 2) at least one specific example (eg. Airflow for `SelfOrchestratedEngine`.

File: src/golang/lib/workflow/engine/aq_engine.go [hunk 13]
```diff
@@ -0,0 +1,811 @@
...
+
+	start := time.Now()
+
+	for len(inProgressOps) > 0 {
+		if time.Since(start) > timeConfig.ExecTimeout {
+			return errors.New("Reached timeout waiting for workflow to complete.")
+		}
+
+		for _, op := range inProgressOps {
+			execState, err := op.GetExecState(ctx)
```
Comment (kenxu95) [inline src/golang/lib/workflow/engine/aq_engine.go:656]: For `op.GetExecState()` and and `op.Launch()`, which doesn't exist but is introduced in https://github.com/aqueducthq/aqueduct/pull/296, should we pass the job manager into them as arguments, or continue to keep the job manager in the op constructors.

Comment (hsubbaraj-spiral) [reply src/golang/lib/workflow/engine/aq_engine.go:657]: We can keep them in the op constructors.

File: src/golang/cmd/server/handler/refresh_workflow.go [hunk 14]
```diff
@@ -79,44 +74,23 @@ func (h *RefreshWorkflowHandler) Prepare(r *http.Request) (interface{}, int, err
...
-	jobSpec := job.NewWorkflowSpec(
-		workflowObject.Name,
-		workflowObject.Id.String(),
-		h.Database.Config(),
-		h.Vault.Config(),
-		h.JobManager.Config(),
-		h.GithubManager.Config(),
+	executeContext, _ := context.WithTimeout(context.Background(), timeConfig.ExecTimeout)
+	//nolint:errcheck
+	go h.Engine.ExecuteWorkflow(
```
Comment (saurav-c) [inline src/golang/cmd/server/handler/refresh_workflow.go:88]: Should the error at least be logged somewhere?

**Review by saurav-c — APPROVED** [review]
As long as the integration tests are passing, looks good to me! Left some minor comments

**Pull request merged**